bpaotu/bpaotu/views.py

```diff
@@ -593,3 +613,115 @@ def dev_only_ckan_check_permissions(request):
     response = '||'.join([digest, data])
 
     return HttpResponse(response)
+
+
+# --------------------------------------------------------------------------------
+# Add Images to Map
+# --------------------------------------------------------------------------------
+
+# img file extension corresponding to pillow processing format and HttpResponse format
+
+IMG_EXTENSION_TABLE = {
+    'jpg': ['JPEG', 'image/jpeg']
+}
+
+
+LOOKUP_TABLE_KEY = 'lkup_tbl_key'
+
+
+def _get_cached_item(key):
+    try:
+        cache = caches['image_results']
+
+        return cache.get(key)
+    except Exception as e:
+        logger.error(str(e))
+
+
+def _set_cached_item(key, value):
+    try:
+        cache = caches['image_results']
+        cache.set(key, value, CACHE_7DAYS)
+    except Exception as e:
+        logger.error(str(e))
+
+    return True
+
+
+def _create_img_lookup_table():
+    '''
+    Create a lookup table of all images in ckan.
+    '''
+    if _get_cached_item(LOOKUP_TABLE_KEY) is None:
+        remote = ckanapi.RemoteCKAN(settings.BPA_PROD_URL, apikey=settings.CKAN_API_KEY)
+
+        packages = remote.action.package_search(
+            fq='tags:site-images',
+            include_private=True,
+            rows=100000
+        )['results']
+
+        logger.info('There are {} image packages.'.format(len(packages)))
+
+        lookup_table = defaultdict(list)
+
+        for i in packages:
+            try:
+                coords = (i['latitude'], i['longitude'])
+                img_url = i['resources'][0]['url']
+
+                lookup_table[coords].append(img_url)
+
+            except Exception as e:
+                logger.error("Either latitude or longitude is missing for: {}".format(img_url))
+                logger.error("Missing parameter is: {}".format(str(e)))
+
+                continue
+
+        lookup_table = dict(lookup_table)
+        _set_cached_item(LOOKUP_TABLE_KEY, lookup_table)
+
+
+def process_img(request=None, lat=None, lng=None, index=None):
+    '''
+    Return specified cached image or fetch image from ckan and resize before caching and returning.
+    '''
+    _create_img_lookup_table()
+
+    lookup_table = _get_cached_item(LOOKUP_TABLE_KEY)
+
+    img_url = lookup_table[(lat, lng)][int(index)]
```

Comment:
what's the `index`? is it sample ID or is something a bit random, which might change over time? I think it'd be better to make it sample Id, and if that's what it is then maybe make that a bit more obvious in the code

Reply:
This is the index to the position in the list containing the image urls. 

This was also kind of @sztamas idea that since we are going to display all the images that match the lat, lng coordinate then we just loop through (kind of like enumerate() function) the list and use the enumerated value to access each image in turn - which is the index here... if that makes sense?

Reply:
OK, but we don't have any stability in the order or length of that list of images in the `lookup_table`, so the problem is that when we ask for `index 1` that hasn't got any long-term semantic meaning - the image we're asking for might not be the image we get back.